sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisServiceImpl.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.redis;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.Set;
+
+import org.apache.beam.sdk.values.KV;
+
+import redis.clients.jedis.Jedis;
+
+/**
+ * An implementation of the {@link RedisService} that actually communicates with a Redis server.
+ */
+public class RedisServiceImpl implements RedisService {
+
+  private final RedisConnection connection;
+
+  public RedisServiceImpl(RedisConnection connection) {
+    this.connection = connection;
+  }
+
+  private class RedisReaderImpl implements Reader {
+
+    private final String keyPattern;
+    private Jedis jedis;
+    private Iterator<String> keysIterator;
+    private KV<String, String> current;
+
+    public RedisReaderImpl(String keyPattern) {
+      this.keyPattern = keyPattern;
+    }
+
+    @Override
+    public boolean start() {
+      jedis = connection.connect();
+      Set<String> keys = jedis.keys(keyPattern);
```

Comment:
The Jedis documentation specifically prohibits using this method.

http://javadox.com/redis.clients/jedis/2.2.0/redis/clients/jedis/Jedis.html#keys(java.lang.String)

Is there a better API you can provide?